### ISSUE TITLE
Deduplication and Maildir use case

### ISSUE THREAD

**Issue by rmoriz** [restic/restic#1097]
using 0.7.0

I want to use restic to backup my [Maildir](https://en.wikipedia.org/wiki/Maildir) to s3, however by design Maildir contains at least one file per mail, many mails are quite small but in total we are around 30GiB. 

As far as I understand http://restic.readthedocs.io/en/latest/design.html?highlight=deduplication#backups-and-deduplication deduplication does not work for those small files, but shoudn't restic at least not re-upload all files on every backup run?

```shell

# first backup
root@mail100:~# restic -r s3:s3.amazonaws.com/my-bucket/restic-mail100 -p /etc/default/restic-password backup   /data
scan [/data]
scanned 255 directories, 1135361 files in 0:31
error for home/Maildir/cur/1499895884.13248_0.mail100:2,fp: Lstat: lstat /data/Maildir/cur/1499895884.13248_0.mail100:2,fp: no such file or directory
warning for /data/Maildir/dovecot.index.cache: expected 426633908 bytes, saved 426764732 bytes
[39:18] 100.00%  12.319 MiB/s  28.367 GiB / 28.356 GiB  1135622 / 1135616 items  1 errors  ETA 0:00 
duration: 39:18, 12.31MiB/s
snapshot ace96fe7 saved


# subsequent backup
root@mail100:~# restic -r s3:s3.amazonaws.com/my-bucket/restic-mail100  -p /etc/default/restic-password backup   /data
using parent snapshot ace96fe7
scan [/data]
scanned 255 directories, 1135374 files in 0:24
[16:26] 34.07%  10.022 MiB/s  9.665 GiB / 28.368 GiB  573890 / 1135629 items  0 errors  ETA 31:52 

# (still uploading)
```

**Comment by fd0** [comment]
Hey, thanks for asking. Deduplication will work just fine for this use case. Restic only uploads the contents of the files once, but since there won't be much duplicate data between different files (emails), the first backup will be as large as data on disk is. For a subsequent backup with just a few mails added here and there, restic will upload only the new data.

It's a different thing for the metadata: Maildirs (especially with large mailboxes) tend to have many files in a directory. For each such directory, restic uploads the metadata again when a single files or timestamp has changed.

In addition, we don't have a cache for the metadata yet, sofor a subsequent backup it is fetched on demand. We'll get to that eventually.

You can try running `restic backup -f`, which does not load any file/dir metadata from the repo, but will re-read all files locally and upload the changed data only. That may be faster for your use case.

I'm going to close this issue for now, please feel free to add  further comments. Thanks!

**Comment by rmoriz** [comment]
I would suggest adding a counter that counts the network traffic occured on subsequent snapshots (or somehow the size of the delta).

I'm also having a crash right at the end of a snapshot which is _probably_ caused by the large JSON index parsing issue as described in #526

**Comment by fd0** [comment]
You're right, that's a not yet solved problem. Sorry.

**Comment by rmoriz** [comment]
looks like #1584 solved the issue for me (using 0.8.3) - and it's also a lot faster now 🎉 👍

**Comment by fd0** [comment]
Nice, thanks for the feedback!
